// gavel/lib/validators/headers-json-example.js
// Generated by CoffeeScript 1.6.3
var HeadersJsonExample, JsonSchema, SchemaGenerator, SchemaProperties, errors, jsonPointer, _ref;

errors = require('../errors');

JsonSchema = require('./json-schema').JsonSchema;

_ref = require('../utils/schema-generator'), SchemaGenerator = _ref.SchemaGenerator, SchemaProperties = _ref.SchemaProperties;

jsonPointer = require('json-pointer');

HeadersJsonExample = (function() {
  function HeadersJsonExample(real, expected) {
    var error, outError,
      _this = this;
    this.real = real;
    this.expected = expected;
    if (typeof this.real !== 'object') {
      throw new errors.MalformedDataError("Real is not an Object");
    }
    if (typeof this.expected !== 'object') {
      throw new errors.MalformedDataError("Expected is not an Object");
    }
    try {
      this.expected = this.prepareHeaders(JSON.parse(JSON.stringify(this.expected)));
    } catch (_error) {
      error = _error;
      outError = new errors.MalformedDataError("Headers validator - Expected malformed:" + error.message);
      outError['data'] = this.expected;
      throw outError;
    }
    try {
      this.real = this.prepareHeaders(JSON.parse(JSON.stringify(this.real)));
    } catch (_error) {
      error = _error;
      outError = new errors.MalformedDataError("Headers validator - Real malformed:" + error.message);
      outError['data'] = this.real;
      throw outError;
    }
    this.schema = this.getSchema(this.prepareHeaders(this.expected));
    if (this.schema !== void 0) {
      if (this.schema['properties'] !== void 0) {
        ['date', 'expires'].forEach(function(header) {
          if (_this.schema['properties'][header] !== void 0) {
            return delete _this.schema['properties'][header]['enum'];
          }
        });
      }
    }
    this.validator = new JsonSchema(this.real, this.schema);
  }

  HeadersJsonExample.prototype.validate = function() {
    return this.output = this.validator.validate();
  };

  HeadersJsonExample.evaluateOutputToResults = function(data) {
    var indexes, results, _i, _ref1, _results;
    results = [];
    if (data === null) {
      return results;
    }
    if (data.length > 0) {
      indexes = (function() {
        _results = [];
        for (var _i = 0, _ref1 = data.length - 1; 0 <= _ref1 ? _i <= _ref1 : _i >= _ref1; 0 <= _ref1 ? _i++ : _i--){ _results.push(_i); }
        return _results;
      }).apply(this);
      indexes.forEach(function(index) {
        var item, message;
        item = data[index];
        console.error;
        message = {
          pointer: jsonPointer.compile(item['property']),
          severity: 'error',
          message: item.message
        };
        return results.push(message);
      });
    }
    return results;
  };

  HeadersJsonExample.prototype.prepareHeaders = function(headers) {
    var key, transformedHeaders, value;
    if (!(headers instanceof Object)) {
      return headers;
    }
    transformedHeaders = {};
    for (key in headers) {
      value = headers[key];
      transformedHeaders[key.toLowerCase()] = value;
    }
    return transformedHeaders;
  };

  HeadersJsonExample.prototype.getSchema = function(data) {
    var properties, schemaGenerator;
    properties = new SchemaProperties({});
    properties.set({
      keysStrict: false,
      valuesStrict: true,
      typesStrict: false
    });
    schemaGenerator = new SchemaGenerator({
      json: data,
      properties: properties
    });
    return schemaGenerator.generate();
  };

  return HeadersJsonExample;

})();

module.exports = {
  HeadersJsonExample: HeadersJsonExample
};
